appclient/src/main/java/org/jboss/as/appclient/service/LazyConnectionContextSelector.java

```diff
@@ -99,14 +101,19 @@ private synchronized void createConnection() {
 
     @Override
     public EJBClientContext getCurrent() {
-        if (clientContext == null) {
-            synchronized (this) {
+        return WildFlySecurityManager.doUnchecked(new PrivilegedAction<EJBClientContext>() {
```

Comment:
doUnchecked() should be as fast as checking yourself (in the latest version of the sec mgr), with the exception of allocating the PrivilegedAction per call (and one interface method call, which should not show up on any realistic execution profile of EJB code).  If you can push the context information into a parameter, and it looks like you can, you can "hoist" that up to a final field or a static final field and use ParametricPrivilegedAction instead, eliminating the allocation overhead as well.